codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -346,20 +580,121 @@ public void onHeadersRead(ChannelHandlerContext ctx, int streamId,
         @Override
         public void onHeadersRead(ChannelHandlerContext ctx, int streamId, Http2Headers headers,
                                   int padding, boolean endOfStream) {
-            Http2HeadersFrame headersFrame = new DefaultHttp2HeadersFrame(headers, endOfStream, padding);
-            headersFrame.streamId(streamId);
-            ctx.fireChannelRead(headersFrame);
+            ctx.fireChannelRead(new DefaultHttp2HeadersFrame(headers, endOfStream, padding)
+                                        .stream(requireStream(streamId)));
         }
 
         @Override
         public int onDataRead(ChannelHandlerContext ctx, int streamId, ByteBuf data, int padding,
                               boolean endOfStream) {
-            Http2DataFrame dataFrame = new DefaultHttp2DataFrame(data.retain(), endOfStream, padding);
-            dataFrame.streamId(streamId);
-            ctx.fireChannelRead(dataFrame);
-
-            // We return the bytes in bytesConsumed() once the stream channel consumed the bytes.
+            ctx.fireChannelRead(new DefaultHttp2DataFrame(data.retain(), endOfStream, padding)
+                                        .stream(requireStream(streamId)));
+            // We return the bytes in consumeBytes() once the stream channel consumed the bytes.
             return 0;
         }
+
+        private <V> Http2Stream2 requireStream(int streamId) {
+            Http2Stream2 stream = connection().stream(streamId).getProperty(streamKey);
+            if (stream == null) {
+                throw new IllegalStateException("Stream object required for identifier: " + streamId);
+            }
+            return stream;
+        }
+    }
+
+    /**
+     * {@link Http2Stream2} implementation.
+     */
+    // TODO(buchgr): Merge Http2Stream2 and Http2Stream.
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)